### PR TITLE
Citylist scene using MVVM design pattern

### PR DESCRIPTION
**Pull request by vandermesis**

### CODE REVIEW

File: CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+//  Copyright © 2020 vandermesis. All rights reserved.
+//
+
+import UIKit
+
+protocol CityListCellManagerMVVMDelegate: class {
+    func didSelectCityCell(city: CityWeather)
+}
+
+final class CityListCellManagerMVVM: NSObject {
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift]: Do you have to use nsobject?

File: CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+}
+
+extension CityListCellManagerMVVM {
+
+    func setup(with citiesWeather: [CityWeather]) {
+        self.citiesWeather = citiesWeather
+    }
+}
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift:28]: In most of MVVM Modells cell manager is not exactly working like that.

Comment (SredzinskiPatryk) [reply CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift:28]: It’s fine and we can leave it as it, but I will also tell you how it looks like in standard products. Już should keep your data souls in view model and Ask cell manager only for cells. Most of them would only have only one method called build cell

File: CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+//  Copyright © 2020 vandermesis. All rights reserved.
+//
+
+import UIKit
+
+final class CityListControllerMVVM: SharedViewController {
+
+    @IBOutlet private weak var cityListTableView: UITableView!
+
+    private var citiesWeatherDataSource = [CityWeather]()
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift]: Keep the data source in view model that’s the whole idea of MVVM

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

File: CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift (hunk 1 above)
Comment (vandermesis) [reply CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift]: If I don't use it here then I get error
`Type 'CityListCellManagerMVVM' does not conform to protocol 'NSObjectProtocol'`
And to fix that I have to import 13 methods:
```
func isEqual(_ object: Any?) -> Bool {
        <#code#>
    }
    
    var hash: Int {
        <#code#>
    }
    
    var superclass: AnyClass? {
        <#code#>
    }

    func `self`() -> Self {
        <#code#>
    }

    func perform(_ aSelector: Selector!) -> Unmanaged<AnyObject>! {
        <#code#>
    }

    func perform(_ aSelector: Selector!, with object: Any!) -> Unmanaged<AnyObject>! {
        <#code#>
    }

    func perform(_ aSelector: Selector!, with object1: Any!, with object2: Any!) -> Unmanaged<AnyObject>! {
        <#code#>
    }

    func isProxy() -> Bool {
        <#code#>
    }

    func isKind(of aClass: AnyClass) -> Bool {
        <#code#>
    }

    func isMember(of aClass: AnyClass) -> Bool {
        <#code#>
    }

    func conforms(to aProtocol: Protocol) -> Bool {
        <#code#>
    }

    func responds(to aSelector: Selector!) -> Bool {
        <#code#>
    }

    var description: String {
        <#code#>
    }
```

File: CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift (hunk 3 above)
Comment (vandermesis) [reply CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift]: I didn't notice that it is not needed here. That var is not used anywhere. I just forgot to delete it after refactoring. :/

Comment (vandermesis) [reply CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift]: But I get a point about not using data source in controller but only in viewmodel. :) I didn't know that.

File: CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+//
+//  Created by Marek Skrzelowski on 11/01/2020.
+//  Copyright © 2020 vandermesis. All rights reserved.
+//
+
+import UIKit
+
+final class CityListCellManagerMVVM: NSObject {
+
+    var citiesWeather = [CityWeather]()
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift]: get rid of this as well, pass array and index in build Cell Methods

File: CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+extension CityListCellManagerMVVM {
+    
+    func buildCell(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeue(with: CityListTableViewCell.self, for: indexPath)
+        cell.setup(with: citiesWeatherDisplayable[indexPath.row])
+        return cell
+    }
+
+    func didSelectCityCell(indexPath: IndexPath) -> CityWeather {
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift]: you can keep it only in controller, no need to continue asking for cell manager in controllers tableview delegate

File: CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        return citiesWeather[indexPath.row]
+    }
+
+    func setupCellManager(with citiesWeather: [CityWeather]) {
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift]: no need this

File: CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    @IBAction private func addButtonPressed(_ sender: UIBarButtonItem) {
+        let favourtieCitiesController = FavouriteCitiesCreator().getController()
+        navigationController?.pushViewController(favourtieCitiesController, animated: true)
+    }
+}
+
+extension CityListControllerMVVM: UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        cellManager.citiesWeather.count
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift]: viewModel.citiesWeather.count

File: CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        return cellManager.buildCell(tableView, cellForRowAt: indexPath)
+    }
+}
+
+extension CityListControllerMVVM: UITableViewDelegate {
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let city = cellManager.didSelectCityCell(indexPath: indexPath)
+        navigateToCityForecast(city: city)
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift:65]: let city = viewModel.citiesWeather[indexPath.row]

Comment (SredzinskiPatryk) [reply CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift:65]: no need cell manager

File: CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    }
+}
+
+extension CityListControllerMVVM: CityListViewModelMVVMDelegate {
+
+    func didUpdateFavouriteCitiesWeather(citiesWeather: [CityWeather]) {
+        toogleSpinner(false)
+        cellManager.setupCellManager(with: citiesWeather)
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift]: no need this

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

File: CleanWeather/Scenes/CityListMVVM/CityListViewModelMVVM.swift [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    }
+}
+
+extension CityListViewModelMVVM {
+
+    func getFavouriteCities() {
+        repository.fetchFavouriteCities { result in
+            switch result {
+            case .success(let cities):
+                self.fetchCitiesWeather(cities: cities) { result in
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListViewModelMVVM.swift]: put it into separated method

File: CleanWeather/Scenes/CityListMVVM/CityListViewModelMVVM.swift [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        let citiesWeatherListResult = convertToCityWeatherListResult(responses: responses)
+        completion?(citiesWeatherListResult)
+    }
+
+    private func convertToCityWeatherResult(city: City, response: Result<CityListAPIResponse, Error>) -> Result<CityWeather, Error> {
+        switch response {
+        case .failure(let error):
+            return .failure(error)
+        case .success(let apiResponse):
+            guard let temp = apiResponse.currently.temperature else {
```
Comment (SredzinskiPatryk) [inline CleanWeather/Scenes/CityListMVVM/CityListViewModelMVVM.swift]: put it into separated method

**Review by SredzinskiPatryk — CHANGES_REQUESTED** [review]

**Review by SredzinskiPatryk — APPROVED** [review]

**Pull request merged**